### PR TITLE
Removes MergeTransitiveOrigin method. Adds null guards to TransitiveOriginsCache

### PR DESCRIPTION
**Pull request by dominoFire**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1645

Regression? Last working version: N/A

## Description

- Fixes exceptions thrown in stack trace called 455b1e49 ; see linked issue. Other call stacks are caused by concurrency issues investigated in another issue.
- Removes MergeTransitiveOrigins method
- Fixes null reference exception thrown in `PackageReferenceMergeComparer`.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added: new unit tests to verify null reference cases in PackageReferenceMergeComparer
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A: No product changes

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs [hunk 1]
```diff
@@ -167,5 +174,58 @@ private static IReadOnlyList<PackageDependency> GetTransitivePackagesForLibrary(
...
+            transitiveEntry?.Keys?.ForEach(fwRuntimePair =>
+            {
+                if (fwRuntimePair != null)
+                {
+                    IList<PackageReference> fwTransitiveOrigins = transitiveEntry[fwRuntimePair];
+                    if (fwTransitiveOrigins != null)
+                    {
+                        foreach (PackageReference transitiveOrigin in fwTransitiveOrigins)
+                        {
+                            if (transitiveOrigin != null && transitiveOrigin.PackageIdentity != null && transitiveOrigin.PackageIdentity.Id != null && transitiveOrigin.PackageIdentity.Version != null)
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]:
```suggestion
                            if (transitiveOrigin?.PackageIdentity != null && transitiveOrigin.PackageIdentity.Id != null && transitiveOrigin.PackageIdentity.Version != null)
```

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs [hunk 2]
```diff
@@ -167,5 +174,58 @@ private static IReadOnlyList<PackageDependency> GetTransitivePackagesForLibrary(
...
+                merged = transitiveOrigins
+                // unique packageId per project
+                .GroupBy(tr => tr.PackageIdentity.Id)
+                // highest version found
+                .Select(g => g.OrderByDescending(pr => pr.PackageIdentity.Version).First())
+                .ToList();
+            }
+            else
+            {
+                merged = new();
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: Do we have to create instance of list?
Could simply return null then check it before use next time?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs [hunk 3]
```diff
@@ -408,25 +405,6 @@ private void MarkTransitiveOrigin(PackageReference top, PackageReference current
...
 
-        internal static TransitivePackageReference MergeTransitiveOrigin(PackageReference currentPackage, TransitiveEntry transitiveEntry)
-        {
-            var transitiveOrigins = new SortedSet<PackageReference>(PackageReferenceMergeComparer);
-
-            transitiveEntry?.Keys?.ForEach(key => transitiveOrigins.AddRange(transitiveEntry[key]));
-
-            List<PackageReference> merged = transitiveOrigins
-                .GroupBy(tr => tr.PackageIdentity.Id)
-                .Select(g => g.OrderByDescending(pr => pr.PackageIdentity.Version).First())
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: Was exception coming from here before your change?

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: any of the members accessed on this LINQ query clause may be null. Now, there are null checks that guarantee that those values are not null.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs (hunk 2 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: > Do we have to create instance of list?

Yes, because there is no null checks on the caller. 

> Could simply return null then check it before use next time?

I prefer empty objects instead of null. I hope the compiler optimizes this.

Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: If this list is not being added to later, use Enumerable.Empty instead of allocating a new list.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs [hunk 4]
```diff
@@ -4,16 +4,23 @@
...
 using System.Linq;
+using Microsoft.VisualStudio.Services.Common;
 using NuGet.Frameworks;
 using NuGet.LibraryModel;
+using NuGet.Packaging;
 using NuGet.Packaging.Core;
 using NuGet.ProjectModel;
 using NuGet.Versioning;
+using NuGet.VisualStudio.Internal.Contracts;
+using TransitiveEntry = System.Collections.Generic.IDictionary<NuGet.Frameworks.FrameworkRuntimePair, System.Collections.Generic.IList<NuGet.Packaging.PackageReference>>;
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: This 'using' obscures the underlying data type and makes the code harder to read. Can we just use Dictionary<FrameworkRuntimePair, List<PackageReferenceProject>> where needed instead? If others disagree, then feel free keep this as-is but I feel pretty strongly that a 'using' pointing to a type (rather than just a namespace) makes the code harder to read.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs [hunk 5]
```diff
@@ -4,16 +4,23 @@
...
+
 
 namespace NuGet.PackageManagement.VisualStudio.Utility
 {
     internal static class GetPackageReferenceUtility
     {
+        internal static readonly Comparer<PackageReference> PackageReferenceMergeComparer = Comparer<PackageReference>.Create((a, b) => a?.PackageIdentity?.CompareTo(b.PackageIdentity) ?? 1);
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: We should have a unit test for this comparer.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs [hunk 6]
```diff
@@ -167,5 +174,58 @@ private static IReadOnlyList<PackageDependency> GetTransitivePackagesForLibrary(
...
+        internal static TransitivePackageReference MergeTransitiveOrigin(PackageReference currentPackage, TransitiveEntry transitiveEntry)
+        {
+            if (currentPackage == null)
+            {
+                throw new ArgumentNullException(nameof(currentPackage));
+            }
+            if (transitiveEntry == null)
+            {
+                throw new ArgumentNullException(nameof(transitiveEntry));
+            }
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]:
```suggestion
            Requires.NotNull(currentPackage, nameof(currentPackage));
            Requires.NotNull(transitiveEntry, nameof(transitiveEntry));
```

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs [hunk 7]
```diff
@@ -167,5 +174,58 @@ private static IReadOnlyList<PackageDependency> GetTransitivePackagesForLibrary(
...
+
+            var transitiveOrigins = new SortedSet<PackageReference>(PackageReferenceMergeComparer);
+            transitiveEntry?.Keys?.ForEach(fwRuntimePair =>
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: If you throw a null argument exception above when transitiveEntry is null, you don't need any '?.' operators when dereferencing transitiveEntry. It's impossible to reach this code if transitiveEntry is null.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs [hunk 8]
```diff
@@ -167,5 +174,58 @@ private static IReadOnlyList<PackageDependency> GetTransitivePackagesForLibrary(
...
+            {
+                if (fwRuntimePair != null)
+                {
+                    IList<PackageReference> fwTransitiveOrigins = transitiveEntry[fwRuntimePair];
+                    if (fwTransitiveOrigins != null)
+                    {
+                        foreach (PackageReference transitiveOrigin in fwTransitiveOrigins)
+                        {
+                            if (transitiveOrigin?.PackageIdentity != null && transitiveOrigin.PackageIdentity.Id != null && transitiveOrigin.PackageIdentity.Version != null)
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs:212]: What is the scenario where the transitiveEntry[fwRuntimePair] would have a null item and therefore fwTransitiveOrigins would contain a null PackageReference that we're calling transitiveOrigin? It's good to have the null check, but is this a valid scenario and if so, are we testing that this doesn't actually occur?

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+using System.Linq;
+using Moq;
+using NuGet.Frameworks;
+using NuGet.PackageManagement.VisualStudio.Utility;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Versioning;
+using NuGet.VisualStudio.Internal.Contracts;
+using Xunit;
+using TransitiveEntry = System.Collections.Generic.IDictionary<NuGet.Frameworks.FrameworkRuntimePair, System.Collections.Generic.IList<NuGet.Packaging.PackageReference>>;
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: This is mostly unused. See comment above about using that obscures data type. I'd recommend removing this in favor of using the actual data type used for consistency with the rest of the test and readability.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+{
+    public class GetPackageReferenceUtilityTests
+    {
+        [Fact]
+        public void MergeTransitiveOrigin_DuplicateTransitiveOrigins_Merges()
+        {
+            // Arrange
+            var net472 = NuGetFramework.Parse("net472");
+            var pr = new PackageReference(new PackageIdentity("packageA", new NuGetVersion("1.0.0")), net472);
+            var te = new Dictionary<FrameworkRuntimePair, IList<PackageReference>>
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]:
```suggestion
            var transitiveEntry = new Dictionary<FrameworkRuntimePair, IList<PackageReference>>
```

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+namespace NuGet.PackageManagement.VisualStudio.Test
+{
+    public class GetPackageReferenceUtilityTests
+    {
+        [Fact]
+        public void MergeTransitiveOrigin_DuplicateTransitiveOrigins_Merges()
+        {
+            // Arrange
+            var net472 = NuGetFramework.Parse("net472");
+            var pr = new PackageReference(new PackageIdentity("packageA", new NuGetVersion("1.0.0")), net472);
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]:
```suggestion
            var packageReference = new PackageReference(new PackageIdentity("packageA", new NuGetVersion("1.0.0")), net472);
```

Other names: pkgReference, pkgRef, etc. A more descriptive variable would make this easier to read.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+            var transitiveOrigin = transitivePackageReference.TransitiveOrigins.Single();
+            Assert.Equal(NuGetVersion.Parse("0.0.2"), transitiveOrigin.PackageIdentity.Version);
+        }
+
+        [Fact]
+        public void MergeTransitiveOrigin_EmptyList_Succeeds()
+        {
+            // Arrange
+            var framework = NuGetFramework.Parse("net6.0");
+            var pr = new PackageReference(new PackageIdentity("packageA", new NuGetVersion("1.0.0")), framework);
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: More descriptive variable name for "pr"

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+            // Act
+            TransitivePackageReference transitivePackageReference = GetPackageReferenceUtility.MergeTransitiveOrigin(pr, te);
+
+            // Assert
+            var transitiveOrigin = transitivePackageReference.TransitiveOrigins.Single();
+            Assert.Equal(NuGetVersion.Parse("0.0.2"), transitiveOrigin.PackageIdentity.Version);
+        }
+
+        [Fact]
+        public void MergeTransitiveOrigin_EmptyList_Succeeds()
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: What does it mean for this to succeed? Can you make the last section of the test method name describe what is expected?
See: https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#unit-test-method-naming

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+            TransitivePackageReference transitivePackageReference = GetPackageReferenceUtility.MergeTransitiveOrigin(pr, transitiveEntry);
+
+            // Assert
+            Assert.Equal(pr.PackageIdentity, transitivePackageReference.PackageIdentity);
+            Assert.Empty(transitivePackageReference.TransitiveOrigins);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetDataWithNulls))]
+        public void MergeTransitiveOrigin_ListWithNulls_Succeeds(List<PackageReference> transitiveOrigins, int expectedElementCount)
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: What does it mean for this to succeed? Can you make the last section of the test method name describe what is expected?
See: https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#unit-test-method-naming

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+                    new PackageReference(new PackageIdentity("package1", new NuGetVersion("0.0.2")), fwRidNetFx.Framework),
+                    null,
+                },
+            };
+
+            // Act
+            TransitivePackageReference transitivePackageReference = GetPackageReferenceUtility.MergeTransitiveOrigin(pr, transitiveEntry);
+
+            // Assert
+            // Target framework does not matter because PM UI doesn't support multitargeting
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]:
```suggestion
            // Target framework does not matter because PM UI doesn't support multi-targeting
```

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+
+        [Fact]
+        public void MergeTransitiveOrigin_NullArgument_Throws()
+        {
+            Assert.Throws<ArgumentNullException>(() => GetPackageReferenceUtility.MergeTransitiveOrigin(null, It.IsAny<TransitiveEntry>()));
+            Assert.Throws<ArgumentNullException>(() => GetPackageReferenceUtility.MergeTransitiveOrigin(It.IsAny<PackageReference>(), null));
+        }
+
+        [Fact]
+        public void MergeTransitiveOrigin_WithNullTransitiveEntryList_ReturnsEmpty()
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: This test method suggests that MergeTransitiveOrigin should be returning Empty but the assertion shows Transitive Origins are a collection of 3 items. Can you fix the method name to describe what's expected?

**Review by jebriede — CHANGES_REQUESTED** [review]
Left a few comments that relate to requested changes and some optional changes. Please let me know if you have any questions or want to discuss any of the feedback! :-)

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs (hunk 3 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: > Was exception coming from here before your change?

No. Original callstack points to the beginning of the method:

```
NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.<>c.<.cctor>b__49_0(PackageReference,PackageReference)
System.Collections.Generic.SortedSet`1.AddIfNotPresent(T)
Microsoft.VisualStudio.Services.Common.CollectionsExtensions.AddRange[T,TCollection](TCollection,IEnumerable`1)
Microsoft.VisualStudio.Services.Common.EnumerableExtensions.ForEach[T](IEnumerable`1,Action`1)
NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.MergeTransitiveOrigin(PackageReference,IDictionary`2)
System.Linq.Enumerable.<>c__DisplayClass7_0`3.<CombineSelectors>b__0(TSource)
System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
System.Collections.Generic.List`1..ctor(IEnumerable`1)
System.Linq.Enumerable.ToList[TSource](IEnumerable`1)
NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.<GetInstalledAndTransitivePackagesAsync>d__36.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.<GetInstalledPackagesAsync>d__6.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.<GetInstalledPackagesAsync>d__5.MoveNext()
```

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs (hunk 4 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: Ok, I will use `IDictionary<FrameworkRuntimePair, List>` to keep it flexible.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs (hunk 5 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: Added unit tests.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs (hunk 10 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: Renamed all instances.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs (hunk 12 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: Renamed all instances to `testPkgReference`.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs (hunk 13 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: Renamed to `MergeTransitiveOrigin_EmptyTransitiveOriginsList_ReturnsEmptyMergedTransitiveOriginsList`

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs (hunk 11 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: Renamed all instances to `testPkgReference`.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs (hunk 15 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: Fixed. Thanks!

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs (hunk 9 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: Removed.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs (hunk 2 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: Replaced with `Enumerable.Empty<PackageReference>()`

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs (hunk 14 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: Renamed to `MergeTransitiveOrigin_WithTransitiveOriginListWithNulls_ReturnsExpectedTransitiveOriginsElementCount`.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs (hunk 16 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Utility/GetPackageReferenceUtilityTests.cs]: Renamed to `MergeTransitiveOrigin_WithNullTransitiveEntryList_ReturnsMergedTransitiveOrigins`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs (hunk 7 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs]: Changed to `transitiveEntry.Keys.ForEach(fwRuntimePair =>`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs (hunk 3 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: Now, I realized the bug is in the comparer. This is also fixed in this PR.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs (hunk 8 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs:212]: > What is the scenario where the transitiveEntry[fwRuntimePair] would have a null item and therefore fwTransitiveOrigins would contain a null PackageReference that we're calling transitiveOrigin?

This can occur if `installedPackages` contains a null `PackageReference` object:

https://github.com/NuGet/NuGet.Client/blob/d21029641e348e06742f5f563edf084ed811ec3a/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs#L270

> is this a valid scenario ...?

Probably not, because `GetInstalledAndTransitivePackagesAsync()` always return a non-null list, but there's no code that guarantee non-null PackageReference objects.

> if so, are we testing that this doesn't actually occur?

This is tested at `MergeTransitiveOrigin_WithNullTransitiveEntryList_ReturnsMergedTransitiveOrigins`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs (hunk 3 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: My comment is a bit orthogonal to the bug, but the stack trace starts at `NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.GetInstalledPackagesAsync`, which is an API we provide to other VS components (Roslyn and Test Explorer are two we know use it), and this API doesn't return transitive origins. While it's nice it caught the bug, `MergeTransitiveOrigin` is doing work that the extensibility API doesn't need, so I hope it's only happening on the first call after the assets file changes. If `MergeTransitiveOrigin` is being called on every `INuGetProjectService.GetInstalledPackagesAsync` is called, then I think we have a potential perf issue.

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: `GetInstalledAndTransitivePackagesAsync()` computes transitive origins once. That was decided on feature development.

Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: > `GetInstalledAndTransitivePackagesAsync()` computes transitive origins once.

That doesn't appear to be the implementation, unless I'm reading the code wrong. Note that the permalinks I'm using are for the dev branch, not your PR branch.

It looks like this is the block that gets cached values: https://github.com/NuGet/NuGet.Client/blob/3a55f08732acadee70bbf1eacceca56ae8dacb5e/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs#L121-L143

However, `MergeTransitiveOrigin` is called outside this block: https://github.com/NuGet/NuGet.Client/blob/3a55f08732acadee70bbf1eacceca56ae8dacb5e/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs#L173

In other words, `MergeTransitiveOrigin` is called for every transitive package, every time `GetInstalledPackagesAsync` is called. `MergeTransitiveOrigin` also uses a lot of LINQ, including `OrderByDescending` for every transitive package, meaning this implementation makes a lot of allocations and therefore risks causing a lot of GCs.

So, perhaps the transitive origins are cached, but the additional work that's being performed to "merge" the result looks quite high, which is not needed when anything other than PM UI calls `GetInstalledPackages`, which includes restore and the extensibility APIs we provide to other VS components.

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: We discovered a performance issue when computing transitive origins. Bugfix is here.
https://github.com/NuGet/NuGet.Client/pull/4652/files 

I think it's possible to remove this merge process since PM UI does not support multi-targeting and it would be possible to 'merge' when doing the graph walk. But, first I need the bugfix merged.

**Comment by erdembayar** [comment]
@dominoFire 
There is merge conflict because of recent merge.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs (hunk 3 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: Update: we still need to merge final results because in multitargeting scenarios, there are multiple graphs with possible different transitive origins.

I will make sure the 'merge' only happens once and then reuse cached results.

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: Update: we still need to merge final results because in multitargeting scenarios, there are multiple graphs with possible different transitive origins.

I will make sure the 'merge' only happens once and then reuse cached results.

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: > > Was exception coming from here before your change?
> 
> No. Original callstack points to the beginning of the method:
> 
> ```
> NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.<>c.<.cctor>b__49_0(PackageReference,PackageReference)
> System.Collections.Generic.SortedSet`1.AddIfNotPresent(T)
> Microsoft.VisualStudio.Services.Common.CollectionsExtensions.AddRange[T,TCollection](TCollection,IEnumerable`1)
> Microsoft.VisualStudio.Services.Common.EnumerableExtensions.ForEach[T](IEnumerable`1,Action`1)
> NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.MergeTransitiveOrigin(PackageReference,IDictionary`2)
> System.Linq.Enumerable.<>c__DisplayClass7_0`3.<CombineSelectors>b__0(TSource)
> System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
> System.Collections.Generic.List`1..ctor(IEnumerable`1)
> System.Linq.Enumerable.ToList[TSource](IEnumerable`1)
> NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.<GetInstalledAndTransitivePackagesAsync>d__36.MoveNext()
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
> System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
> NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.<GetInstalledPackagesAsync>d__6.MoveNext()
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
> System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
> NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.<GetInstalledPackagesAsync>d__5.MoveNext()
> ```

Where did you above call stack? Could you share me a link? I want to see some other details.

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: @erdembayar Shared via private message.

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: @zivkan I will address performance bug in another issue. See https://github.com/NuGet/Client.Engineering/issues/1675

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:431]: Everyone. I will split this PR into smaller ones to address each issue and call stack error separately.

**Comment by dominoFire** [comment]
I will split this PR into smaller ones to address each issue and call stack error separately.
